### DISCUSSION TITLE
Nuxt Scripts

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#22016] [RFCs]
# Nuxt Scripts RFC

## Summary

The `nuxt/scripts` integration allows third-party scripts to be managed and optimised, following best practices for performant and compliant websites.

Out of the box, Nuxt will provides presets for a number of third-party script providers (such as `gtm`, `crisp`, etc.). Modules can add their own scripts, hooking into the underlying API that `nuxt/scripts` provides, ensuring users benefit from shared best practice.

## Implementation

Users can register third-party scripts either globally or at runtime in a Nuxt plugin or within a Vue component.

When used locally, Nuxt will tree-shake out unused providers from the build, meaning they are strictly opt-in. 

#### Global registration

These scripts will be registered for every page.

```ts
export default defineNuxtConfig({
  scripts: {
    globals: [
      'some-script', // refers to a simple preset if no options are required
      {
        preset: 'gtm', // can be used to refer to a preset that we then override/customise
        strategy: 'partytown' // for example, a custom strategy enabled by the partytown module
      },
      {
        name: 'my-script',
        // custom external script where there is no provider preset
        src: 'https://test.com/example.js',
        // this requires an explicit trigger to register the script and is compatible with
        // registering scripts that depend on cookie consent
        strategy: 'triggered',
      }
    ]
  }
})
```

#### Local use

For full customisation, Nuxt will expose a `useScript` composable that can be called anywhere in the app. Using it will allow granular use of scripts on certain pages. By default that will mean the script is added on the first visit to this page (+ optional setup is called).

```vue
<script setup>
const { status, error } = useScript({
  // all the options of global scripts are supported here
  preset: 'gtm',
  // you can set a unique script name to de-duplicate script loading when multiple pages
  // need access to the same script
  name: 'my-script',
  // as well as a setup function that runs the first time a script is loaded
  setup () {}
})

// it can be awaited if you have code in a setup function that needs to ensure the script is loaded
await useScript(/* */)

// You can also use useScript to trigger loading a script that has previously been registered
// with `strategy: 'triggered'`
useScript('my-script')
</script>
```

**Comment by RodrigoProjects** [comment]
Should `useScript` also provide a loading state that is a `ref` with a false value until `script.onload` triggers? (Useful for client-side)

**Reply by danielroe** [reply]
The `status` ref is envisioned to provide this, with a bit more information than just `load: true/false`.

**Comment by owlnai** [comment]
- How does a 3P qualify to be added in `nuxt/scripts`? Will it be similar to the module listing process in `nuxt/modules` so everyone can contribute? 
- Many external scripts need a bit of inline JS to initialize. Some are static, others are variable (e.g an ID). [Third Party Capital](https://github.com/GoogleChromeLabs/third-party-capital/blob/main/src/third-parties/google-analytics/data.json) approaches this by adding a `code` property. Should we just create a client plugin in that case?

**Reply by danielroe** [reply]
The `setup()` hook is meant to be a good place to initialise the external script.

The aim is where it is appropriate, for third party script presets to be contributed by modules. So, for example, https://github.com/nuxt-modules/turnstile would be able to provide the turnstile integration by hooking into the API provided by `nuxt/scripts`.

**Comment by OhB00** [comment]
CSP support would be a very fitting addition to this.

**Comment by flozero** [comment]
As much as I like the idea people are going to be confused with the modules here. Wouldn't make more sens to have a module scripts ? instead ?

**Comment by ThimoDEV** [comment]
Will this module get all the same base strategy options as the NextJS counterpart?

- beforeInteractive: Load the script before any Next.js (Nuxt) code and before any page hydration occurs.
- afterInteractive: (default) Load the script early but after some hydration on the page occurs.
- lazyOnload: Load the script later during browser idle time.
- worker: (experimental) Load the script in a web worker.

For our clients I was really happy to be able to use lazyonload because the other methods didn't work

**Comment by harlan-zw** [comment]
I have started investigating the implementation of Nuxt Scripts.

This is a draft and will be continuing to work on this

To begin with, we should clarify what exactly we aim to improve.

## Background

Loading third-party scripts in Nuxt is already fairly straight-forward through `useHead` or `nuxt.config`.

```ts
useHead({
  script: [
    { src: 'https://example.com.com/script.js', async: true, defer: true }
  ]
})
```

This works as expected.

We can also make use of [DOM events](https://unhead.harlanzw.com/guide/guides/dom-event-handling) and [Tag Deduping](https://unhead.harlanzw.com/guide/guides/handling-duplicates)
to initialise scripts and make sure it only occurs once.

```ts
useHead({
  script: [
    {
     // ensures only one instance of the script is laoded
     key: 'my-script',
     src: 'https://example.com.com/script.js',
     async: true, 
     defer: true,
     onload() {
        // do something once it's loaded
     }
    }
  ]
}, { mode: 'client' })
```

We also have the option to insert inline scripts.

```ts
useHead({
  script: [
    { innerHTML: 'window.myImportantVar = 1' }
  ]
})
````

## Opportunities

With the above code in mind, we can look at some _possible_ areas that could be improved on the above.

### Loading Strategies

Third-party scripts are usually the main factor left when trying to improve performance, 
and in many cases, it can feel like there's little that can be done.

For example, Analytics Live Chat Widget, Cookie Consent Banner, etc.

We can take inspiration from [next/script](https://nextjs.org/docs/pages/building-your-application/optimizing/scripts),
providing strategies for how scripts are loaded.

While their API is good and established, I would suggest the following options:
- init: Load the script before any Nuxt code and before any page hydration occurs.
- ready: (default) Load the script once Nuxt has finished hydrating, but some scripts may still be loading. (this is the same a `<script async defer>`)
- idle: Load the script once Nuxt is ready and the browser is idle.
- promise: Load the script once a promise is resolved

Support for modules extending the types would be provided, allowing for:

- partytown: (experimental) Load the script in a web worker. (Added by partydown module)

### Asset Strategy

A performance hit with third-party scripts is additional DNS lookups and latency.
Scripts may not be as close as your app.

We can improve this by supporting the strategies from [Nuxt Assets](https://github.com/nuxt/nuxt/discussions/22012).

```ts
useScript({
  src: 'https://example.com.com/script.js',
  assetStrategy: 'proxy'
})
```

By default, no asset strategy would be set and the script would be loaded as-is.

### Security Improvements

In using an `assetStrategy` we also have the opportunity to improve the security of our scripts.
We can automatically append `integrity` and `crossorigin` attributes to the script tag as well as `Content-Security-Policy` headers.

### Improved Module Integrations

Modules will be able to register custom strategies and presets.

## APIs

### `useScript`

The useScript composable is a utility to make loading and using third-party scripts easier.

Following some conventions from [next/script](https://nextjs.org/docs/pages/building-your-application/optimizing/scripts), 

```ts
export default defineComponent({
  setup() {
    const { status, error } = useScript({
      // all the options of global scripts are supported here
      preset: 'gtm',
      // you can set a unique script name to de-duplicate script loading when multiple pages
      // need access to the same script
      name: 'my-script',
      // as well as a setup function that runs the first time a script is loaded
      setup () {}
    })
  }
})
```

### nuxt.config `scripts` - global

The `global` key allows you
to load in popular third-party scripts globally with type-safe options that load in with best practices.

```ts
export default defineNuxtConfig({
    scripts: {
        globals: {
            googleAnalytics: {
                id: '<key>'
            },
            googleTagManager: {
                id: '<id>',
            }
        }
    }
})
```

### nuxt.config `scripts` - presets

The `presets` key allows you to define a named preset of configuration that can be used in `useScript`.

```ts
export default defineNuxtConfig({
    scripts: {
        presets: {
          name: 'my-script',
          // custom external script where there is no provider preset
          src: 'https://test.com/example.js',
          // this requires an explicit trigger to register the script and is compatible with
          // registering scripts that depend on cookie consent
          strategy: 'triggered',
        }
    }
})
```

### nuxt.config `scripts` - defaults

The `defaults` key allows you to define default options for all scripts.

```ts
export default defineNuxtConfig({
    scripts: {
        defaults: {
          // all scripts will be idle loaded by default
          strategy: 'idle',
        }
    }
})
```

### Third Party Capital Integration

We can also expose components generated at runtime using [Third Party Capital](https://github.com/GoogleChromeLabs/third-party-capital).

```vue
<template>
  <GoogleMapsEmbed :id="key" />
</template>
```

**Reply by housseindjirdeh** [reply]
This is great, thanks for sharing @harlan-zw ❤️ 

Out of curiosity, are you thinking of having all of these options provided to developers to choose from? For example—want to load a third-party globally? Use `scripts.globals` in `nuxt.config`. Otherwise, use `useScript`.

Not saying this is a bad idea (it actually sounds pretty great). Just curious :)

**Reply by harlan-zw** [reply]
Thanks :)

This is a tentative idea, I need to explore what global script providers we could actually support and if it's worth the effort. There are advantages to being able to register a script via nuxt.config in that there's no runtime overhead, everything can be SSR.
